File: neptune_py/skeleton/messager.py
import collections
import abc
import struct
from neptune_py.skeleton import utils

NeptuneMessageTuple = collections.namedtuple('NeptuneMessageTuple', ['mtype', 'message'])


class NeptuneMessageType:
    NeptuneMessageTypeNone = 0
    NeptuneMessageTypeNornmal = 1
    NeptuneMessageTypeCall = 2


class NeptuneWriterBaseAbstract:
    @abc.abstractmethod
    def write(self, message):
        raise NotImplementedError()

    @abc.abstractmethod
    def close(self):
        raise NotImplementedError()


class NeptuneMessager:
    MessageTypeFormat = '!H'
    MessageTypeSize = struct.calcsize(MessageTypeFormat)

    def __init__(self, id_, manager, writer: NeptuneWriterBaseAbstract, entity_cls):
        self.manager = manager
        self.entity_cls = entity_cls
        self.id_ = id_
        self.writer = writer
        self.entity = None

    def write_message(self, mtype, message):
        self.writer.write(struct.pack(self.MessageTypeFormat, mtype) + message)

    def on_message(self, message):
        if len(message) < self.MessageTypeSize:
            utils.color_print(utils.AnsiColor.FAIL, "invalid message, cannot unpack message type")
            self.close()
            return
        mtype, *_ = struct.unpack(self.MessageTypeFormat, message[:self.MessageTypeSize])
        self.entity.on_message(NeptuneMessageTuple(mtype=mtype, message=message[self.MessageTypeSize:]))

    def on_connected(self):
        self.entity = self.entity_cls()
        self.entity.bind_messager(self)

    def on_error(self, error):
        raise NotImplementedError()

    def on_disconnected(self):
        self.entity.on_messager_lost(self)

    def close(self):
        self.writer.close()


class NeptuneMessagerManager:
    def __init__(self, entity_cls):
        self.entity_cls = entity_cls
        self.messagers = {}

    def on_connected(self, id_, writer: NeptuneWriterBaseAbstract):
        messager = NeptuneMessager(
            id_, self, writer, self.entity_cls
        )
        messager.on_connected()
        self.messagers[id_] = messager

    def on_message(self, id_, message):
        messager = self.messagers.get(id_)
        if messager is None:
            # TODO invaild, close
            pass

        messager.on_message(message)

    def on_disconnected(self, id_):
        utils.color_print(utils.AnsiColor.FAIL, f'disconnected============{id_}')
        self.remove_messager(id_)

    def remove_messager(self, id_):
        messager = self.messagers.pop(id_, None)
        if messager is not None:
            messager.on_disconnected()

File: lua/neptune/skeleton/messager.lua
local common = require("neptune.skeleton.common")

local exports = {}

exports.NeptuneMessageType = common.Constant({
   NeptuneMessageTypeCall = 1
})


local NeptuneMessager = common.Class.new('NeptuneMessager')
NeptuneMessager.__index = NeptuneMessager
function NeptuneMessager:ctor(writer, entity)
   local o = setmetatable({
         writer = writer,
         entity = entity,
      },
      self
   )
   return o
end

function NeptuneMessager:OnConnected()
   self.entity:BindMessager(self)
end

function NeptuneMessager:OnMessage(mtype, msg)
   if self.entity == nil then
      error("this messager doesnt have entity")
   end
   self.entity:OnMessage(mtype, msg)
end

function NeptuneMessager:OnError(err)
   print("not implemented")
end

function NeptuneMessager:OnDisconnected()
   self.entity:OnMessagerLost()
end

function NeptuneMessager:Close()
   self.writer:Close()
end

function NeptuneMessager:WriteMessage(mtype, msg)
   self.writer:WriteMessage(mtype, msg)
end

-- client-side doesnt need an manager

-- local NeptuneMessagerClientManager = common.Class.new('NeptuneMessagerClientManager')
-- NeptuneMessagerClientManager.__index = NeptuneMessagerClientManager
-- function NeptuneMessagerClientManager:ctor(entity_cls)
--    local o = setmetatable({
--          entity_cls = entity_cls
--          messagers = {}
--       },
--       self
--    )
--    return o
-- end

-- function NeptuneMessagerClientManager:OnConnected(id, read_writer)
--    self.messagers[id] = NeptuneMessager:ctor(id, self, read_writer)
--    self.messagers[id].OnConnected()
-- end

-- function NeptuneMessagerClientManager:OnMessage(id, mtype, msg)
--    local messager = self.messagers[id]
--    if messager == nil then
--       error('unknown message id=' .. id)
--    end
--    messager.OnMessage(id, mtype, msg)
-- end

-- function NeptuneMessagerClientManager:OnDisconnected(id)
--    local messager = self.messagers[id]
--    if messager == nil then
--       error('unknown message id=' .. id)
--    end
--    messager.OnDisconnected()
--    self.messagers[id] = nil
-- end

-- function NeptuneMessagerClientManager:OnError(id, err)
--    local messager = self.messagers[id]
--    if messager == nil then
--       error('unknown message id=' .. id)
--    end
--    messager.OnError(err)
--    self.messagers[id] = nil   
-- end


-- export
exports.NeptuneMessager = NeptuneMessager
-- NeptuneSkeleton.NeptuneMessagerClientManager = NeptuneMessagerClientManager

return exports
